### PR TITLE
add new bq job timeout and retry config

### LINKED ISSUE #45: [Feature] dbt-Bigquery retry for DML job BQ API errors (503 errors, etc. )
### Is there an existing feature request for this?

- [X] I have searched the existing issues

### **Background**

Currently, dbt-bigquery profile.yml has two configurations: [Timeouts](https://docs.getdbt.com/reference/warehouse-profiles/bigquery-profile#timeouts) and [Retries](https://docs.getdbt.com/reference/warehouse-profiles/bigquery-profile#retries)

We at Semios have been using dbt extensively to run high-frequency data pipelines on BigQuery (some of them run every 10 minutes). Because of the strict SLA requirement, we require more granular control on BigQuery timeout and retry mechanism to minimize the intermittent query failures and its impact to our pipeline.  

We found that the existing configs are not enough to mitigate all intermittent query failures. for example, the [Deadline of 120.0s exceeded 503 errors](https://getdbt.slack.com/archives/C99SNSRTK/p1627926805026400) raised by BigQuery JobInsert API that occurred from time to time and impacted several dbt-BigQuery users in the dbt community.

We would like to suggest a more comprehensive BigQuery retry config to minimize all these errors. 

### **Technical Context**

Before I propose the suggestion, let me explain how BigQuery query works and how we would like to retry.

#### How BigQuery query works in dbt
at the core, BigQuery query is made by [two steps in dbt](https://github.com/dbt-labs/dbt-core/blob/3dc574e5ffad53e21bfa352b10bf4ed246dc9f95/plugins/bigquery/dbt/adapters/bigquery/connections.py#L523-L530)
```python
    def _query_and_results(self, client, sql, conn, job_params, timeout=None):
        """Query the client and wait for results."""
        # Cannot reuse job_config if destination is set and ddl is used
        job_config = google.cloud.bigquery.QueryJobConfig(**job_params)    # <--- Step 1
        query_job = client.query(sql, job_config=job_config)               # <--- Step 2
        iterator = query_job.result(timeout=timeout)
        return query_job, iterator
```
In the first step, `client.query()` submits a query to BQ JobInsert API server, when succeeded, BQ server creates a new BigQuery query job, and return the `query job id` back to the client as part of`query_job` object. This step shall be very quick, normally under a few seconds. however, in some rare cases, it would take much longer and might even up to 4 minutes according to the BigQuery engineering team. 

In the 2nd step, `query_job.result()` await for the BigQuery executing (running) the query and return the results back to the client as an iterator. depending on the complexity of the query, this step could takes long, from tens of seconds to tens of minutes. 

#### how dbt currently retries BigQuery queries
Currently, The dbt `retry` config is used to set the retries of the overall `_query_and_results`. the dbt `timeout` config is used only to control step 2, `query_job.result(timeout=timeout)`. These configs make sense on its own. 

However, you might notice that there is no control over the timeout of `query()` in step 1, other than relying on its [default value] (https://github.com/googleapis/python-bigquery/blob/1762e5bd98ab089abf65a0d146234c77217b8dbe/google/cloud/bigquery/client.py#L3193).  When BQ JobInsert API sharding cluster is unstable. it could take up to 4 minutes to create the query job id in some rare cases. Because the `query()` had a client-side [default deadline of 120s](DEFAULT_RETRY = retry.Retry(predicate=_should_retry)), the client quits at 120s while server side is still waiting for the job creation. That's why several dbt community members had experienced the [Deadline of 120.0s exceeded errors (https://getdbt.slack.com/archives/C99SNSRTK/p1627926805026400).

BQ team has since patched the `query()` to make a client-side [default deadline of 10 mins](DEFAULT_RETRY = retry.Retry(predicate=_should_retry, deadline=600.0). It means by default, `query()` could take up to 4 minutes to make one attempt to create a new job id. If it fails, the client-side could have up to 10 minutes to retry. This would work fine for users who don't have stringent timing requirements. They won't mind waiting for 10 minutes to create a new BQ job ID and then waiting for another 10 minutes to get the result. 

Unfortunately, it doesn't work for us. In situations like this, we would rather fail faster and try again. It shall not take more than 30 seconds to create a new job id. if it takes too long, we would rather let the query() timeout and fail, so that we could retry it again. 

### **the Proposed Solution**
In order to gain fine control of the timeout mechanism of step 1 and step 2, we would like to propose the following 4 dbt configs
```python
job_creation_timeout_seconds     # specific for initiate BQ job, to control the timeout of step 1, query()
job_execution_timeout_seconds    # specific for awaiting job result, to control the timeout of step 2, result()

job_retry_deadline_seconds       # to control the overal query, retry_deadline of _query_and_results()
job_retries                      # to control the overall query, retries of _query_and_results()
```

For example, we could set the configs below to fail faster on the step of BQ job creation, while allowing queries with long-running results.
```python
job_creation_timeout_seconds=30
job_execution_timeout_seconds=1200
job_retry_deadline_seconds=1500
job_retries=3
```
These settings would allow us to control the timeout behaviors of step 1 and step 2 on their own, hence maximizing our chances to mitigate different kinds of intermittent errors.

NOTE: 
`job_execution_timeout_seconds` is the renaming of the current `timeout` config.
`job_retries` is the renaming of the current `retries` config.

### Describe alternatives you've considered

Currently, we have to override the dbt-core code to allow the `query()` to fail faster and retry, and to mitigate the 503 error.

### Who will this benefit?

the dbt-bigquery users who would need more fine control of BigQuery query behaviors

### Are you interested in contributing this feature?

yes

### Anything else?

this issue has been discussed by several dbt community members in slack
https://getdbt.slack.com/archives/C99SNSRTK/p1627926805026400

### PR DESCRIPTION
**Pull request by hui-zheng**
resolves #45 

### Description

Description
This PR provides a fine-grained control of the timeout and retry to bq query with four dbt-profile configs

```
job_creation_timeout_seconds     # specific for initiate BQ job, to control the timeout of step 1, query()
job_execution_timeout_seconds    # specific for awaiting job result, to control the timeout of step 2, result()

job_retry_deadline_seconds       # to control the overal query, retry_deadline of _query_and_results()
job_retries                      # to control the overall query, retries of _query_and_results()
```

These settings would allow us to control the timeouts of step 1 and step 2 of the BigQuery on their own ( See below), hence maximizing the chances to mitigate different kinds of intermittent errors.

For example, we could set the configs below to fail faster at the step of BQ job creation, while allowing queries with long-running results.
```
job_creation_timeout_seconds=30
job_execution_timeout_seconds=1200
job_retry_deadline_seconds=1500
job_retries=3
```
NOTE:
job_execution_timeout_seconds is the renaming of the previous timeout config.
job_retries is the renaming of the previous retries config.

Context
---

at the core, BigQuery query is made by two steps in dbt

    def _query_and_results(self, client, sql, conn, job_params, timeout=None):
        """Query the client and wait for results."""
        # Cannot reuse job_config if destination is set and ddl is used
        job_config = google.cloud.bigquery.QueryJobConfig(**job_params)    # <--- Step 1
        query_job = client.query(sql, job_config=job_config)               # <--- Step 2
        iterator = query_job.result(timeout=timeout)
        return query_job, iterator
In the first step, client.query() submits a query to BQ JobInsert API server, when succeeded, BQ server creates a new BigQuery query job, and return the query job id back to the client as part ofquery_job object. This step shall be very quick, normally under a few seconds. however, in some rare cases, it would take much longer and might even up to 4 minutes according to the BigQuery engineering team.

In the 2nd step, query_job.result() await for the BigQuery to execute (running) the query and return the results to the client as an iterator. depending on the complexity of the query, this step could takes long, from tens of seconds to tens of minutes.

--- 

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-bigquery next" section.

### CODE REVIEW

**Comment by hui-zheng** [comment]
hello @jtcohen6,
I submit the first draft of the PR. I am going to add some tests later. It's my first time contributing. I appreciate any feedback and suggestions for the next step.

File: dbt/adapters/bigquery/connections.py [hunk 1]
```diff
@@ -113,6 +116,8 @@ class BigQueryCredentials(Credentials):
         'dataset': 'schema',
         'target_project': 'target_database',
         'target_dataset': 'target_schema',
+        'retries': 'job_retries',
+        'timeout_seconds': 'job_execution_timeout_seconds',
```
Comment (jtcohen6) [inline dbt/adapters/bigquery/connections.py:120]: This is clever! Aliases also affect node-level configs, but I don't see a reason why we can't do it this way

**Review by jtcohen6 — COMMENTED** [review]
@hui-zheng These changes look good to me.

As with all connection-related issues, it's difficult to stand up integration tests that replicate behavior in the wild. If you've been running this branch locally / in your deployment environment with success, that's a pretty strong vote of confidence for me. Is that something you've been able to do?

If so, could you also add an entry to the changelog (`dbt-bigquery 1.0.0 (Release TBD)` > `Features`), and add yourself to the list of contributors?

After merging, the last step is to update the prerelease docs: https://next.docs.getdbt.com/reference/warehouse-profiles/bigquery-profile#optional-configurations

**Comment by hui-zheng** [comment]
@jtcohen6 

I haven't tested out this exact PR in our deployment environment, because We are using dbt v0.20.2, in which the dbt-bigquery is still part of dbt-core.

I am happy to test this PR out. Could you let me know which stable version of dbt-core it could pair up for testing?

**Comment by jtcohen6** [comment]
@hui-zheng This change will be going into dbt-bigquery v1.0.0, which is still in prerelease (and pairs with the dbt-core v1.0.0 prereleases). So there isn't a stable version (yet) to test in production.

By installing this package locally (`pip install -e .`), it will also install the latest compatible prerelease of `dbt-core`, which is `1.0.0-b2`. For the sake of the changes you're making, that should work fine.

As it is, these changes look good. Given that we're coming up on v1.0.0-rc1, we may merge to include and test more in the release candidate.

**Comment by jtcohen6** [comment]
Going to close and reopen just to trigger adapter integration tests

**Comment by hui-zheng** [comment]
@jtcohen6 

I try  `pip install -e .` in `/dbt-bigquery` and got an error, saying
```
    dbt-core 1.0.0b2 depends on dbt-extractor==0.4.0
    dbt-core 1.0.0b1 depends on dbt-extractor==0.4.0
```
My branch is up to date with `dbt-bigquery` main branch.

My python versions
```
❯ python --version
Python 3.8.12
❯ pip --version
pip 21.1.1 from /Users/hui-zheng/.pyenv/versions/3.8.12/envs/dbt-lab/lib/python3.8/site-packages/pip (python 3.8)
```

Do you know how to resolve it?

<details>
  <summary>Full error message, Click to expand!</summary>
  
```
Collecting dbt-extractor==0.4.0
  Using cached dbt_extractor-0.4.0.tar.gz (21 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... error
    ERROR: Command errored out with exit status 1:
     command: /Users/hui-zheng/.pyenv/versions/3.8.12/envs/dbt-lab/bin/python3.8 /Users/hui-zheng/.pyenv/versions/3.8.12/envs/dbt-lab/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /var/folders/g6/3qpc1f9533j6hgcwjzcp__cr0000gn/T/tmp4iygtpl0
         cwd: /private/var/folders/g6/3qpc1f9533j6hgcwjzcp__cr0000gn/T/pip-install-5u3uc0sp/dbt-extractor_4897a4d46c244cb7a180c2507be2db66
    Complete output (6 lines):

    Cargo, the Rust package manager, is not installed or is not on PATH.
    This package requires Rust and Cargo to compile extensions. Install it through
    the system's package manager or via https://rustup.rs/

    Checking for Rust toolchain....
    ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/a7/5c/609f02383178208612d6ac21228ca256337d3c18afb13b29f122720a26ad/dbt_extractor-0.4.0.tar.gz#sha256=58672e36fab988c849a693405920ee18421f27245c48e5f9ecf496369ed31a85 (from https://pypi.org/simple/dbt-extractor/). Command errored out with exit status 1: /Users/hui-zheng/.pyenv/versions/3.8.12/envs/dbt-lab/bin/python3.8 /Users/hui-zheng/.pyenv/versions/3.8.12/envs/dbt-lab/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /var/folders/g6/3qpc1f9533j6hgcwjzcp__cr0000gn/T/tmp4iygtpl0 Check the logs for full command output.
Collecting dbt-core~=1.0.0b1
  Using cached dbt_core-1.0.0b1-py3-none-any.whl (789 kB)
INFO: pip is looking at multiple versions of <Python from Requires-Python> to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of dbt-bigquery to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install dbt-bigquery because these package versions have conflicting dependencies.

The conflict is caused by:
    dbt-core 1.0.0b2 depends on dbt-extractor==0.4.0
    dbt-core 1.0.0b1 depends on dbt-extractor==0.4.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
```
</details>

File: dbt/adapters/bigquery/connections.py [hunk 2]
```diff
@@ -323,17 +330,24 @@ def open(cls, connection):
         return connection
 
     @classmethod
-    def get_timeout(cls, conn):
+    def get_job_execution_timeout_seconds(cls, conn):
```
Comment (jtcohen6) [inline dbt/adapters/bigquery/connections.py:333]: There are some integration test failures like:
```
Unhandled error while executing ...
'BigQueryConnectionManager' object has no attribute 'get_timeout'
```

We need to replace `get_timeout` &rarr; `get_job_execution_timeout_seconds` within `load_dataframe`, to fix `dbt seed`:
https://github.com/dbt-labs/dbt-bigquery/blob/bc6113cc8efbc6a6f4618b5839e5a8873a48d2d2/dbt/adapters/bigquery/impl.py#L700

Plus one unit test:
https://github.com/dbt-labs/dbt-bigquery/blob/bc6113cc8efbc6a6f4618b5839e5a8873a48d2d2/tests/unit/test_bigquery_adapter.py#L603

**Review by jtcohen6 — COMMENTED** [review]
@hui-zheng I don't know exactly what went wrong there, but it seems like an issue with trying to install `dbt-bigquery` in an environment that already has `dbt-core==1.0.0b1`. The simplest solution might be to clear away the virtual environment, create a fresh one, and try again from there.

File: dbt/adapters/bigquery/connections.py (hunk 2 above)
Comment (hui-zheng) [reply dbt/adapters/bigquery/connections.py:333]: I could look into that

**Review by hui-zheng — COMMENTED** [review]
updated `Changelog`

**Comment by hui-zheng** [comment]
> the last step is to update the prerelease docs: https://next.docs.getdbt.com/reference/warehouse-profiles/bigquery-profile#optional-configurations

@jtcohen6 , How do I update the docs?

**Comment by jtcohen6** [comment]
I triggered the integration test to re-run. (We had some intermittent failures over the past few days due to our sandbox BQ environment.)

It looks like there are some flake8 failures (simple), and failures in the unit tests related to mock code (less simple). One of the failures looks like:
```
unsupported type for timedelta seconds component: MagicMock
```
Unfortunately, I can't dig in deeper right now. It should be somewhat easier to run + reproduce reproduce those unit test failures locally. Let me know if you're still stuck later this week, and I can try to dive in.

> How do I update the docs?

Docs updates happen in this repo: https://github.com/dbt-labs/docs.getdbt.com

I recommend opening an issue, and then (if you feel up to it) a PR targeting `next` to resolve it as well. As a shortcut, you can click the "Edit this page" link at the bottom of each page in the docs. Clicking that for the page I linked would take you here: https://github.com/dbt-labs/docs.getdbt.com/edit/next/website/docs/reference/warehouse-profiles/bigquery-profile.md

**Comment by jtcohen6** [comment]
@hui-zheng Sorry about that, I tried merging in the recent changes from `main`, but it looks like I may have added a few `flake8` errors of my own. Feel free to remove those commits from your branch.

There's still a handful of unit tests failing:
```
FAILED tests/unit/test_bigquery_adapter.py::TestBigQueryConnectionManager::test_copy_bq_table_appends
FAILED tests/unit/test_bigquery_adapter.py::TestBigQueryConnectionManager::test_copy_bq_table_truncates
FAILED tests/unit/test_bigquery_adapter.py::TestBigQueryConnectionManager::test_drop_dataset
FAILED tests/unit/test_bigquery_adapter.py::TestBigQueryConnectionManager::test_query_and_results
FAILED tests/unit/test_bigquery_adapter.py::TestBigQueryConnectionManager::test_retry_and_handle
FAILED tests/unit/test_bigquery_adapter.py::TestBigQueryConnectionManager::test_retry_connection_reset
```
The issue is that this PR has changed the function signature of `_query_and_results` (`job_execution_timeout`+ `job_creation_timeout` replaced `timeout`), and the expected return value of `_retry_and_handle` (`deadline` is no longer just `None`). That has implications for these unit tests, which "mock" the anticipated inputs and outputs of those methods. I don't think it would be tremendously difficult to debug those, e.g. to fix `test_query_and_results` I just needed to change [this line](https://github.com/dbt-labs/dbt-bigquery/blob/7e496af5ef17246a379ab14225162819bc5d3069/tests/unit/test_bigquery_adapter.py#L611) to:
```
          query='sql', job_config=mock_bq.QueryJobConfig(), timeout=None)
```

**Comment by hui-zheng** [comment]
> which "mock" the anticipated inputs and outputs of those methods. I don't think it would be tremendously difficult to debug those

@jtcohen6  sorry for the delay in responding to you. a bit busy lately. will fix those testings once I have time

**Comment by hui-zheng** [comment]
@jtcohen6 
I fixed all failed unit tests and passed the local `tox` run. Hope it resolves everything. 

I haven't updated the docs, let me know if it's something I shall do before this PR gets merged or after.

**Comment by hui-zheng** [comment]
@jtcohen6 

I also updated the docs. Please see it here.  https://github.com/dbt-labs/docs.getdbt.com/pull/962

Please let me know if anything else is required.

**Comment by McKnight-42** [comment]
Hi @hui-zheng seems like this branch is falling behind to main and is failing some testing checks now, if you could update it to main would be very much appreciated.

**Comment by hui-zheng** [comment]
Hi @McKnight-42, I just bring it up to date.

**Review by McKnight-42 — APPROVED** [review]
great work on this!

**Pull request merged**